---
bibliography: Supplemental_file.bib
biblio-style: apalike
highlight_bw: yes
output:
  rmarkdown::pdf_document:
    includes:
        in_header: header.tex 
    dev: "cairo_pdf"
    latex_engine: xelatex
    fig_caption: yes
---

**Supplemental File of**

# ggmsa: a visual exploration tool for multiple sequence alignment and associated data

\renewcommand{\figurename}{Fig.}

\newcommand{\beginsupplement}{%
        \setcounter{table}{0}
        \renewcommand{\thetable}{S\arabic{table}}%
        \setcounter{figure}{0}
        \renewcommand{\thefigure}{S\arabic{figure}}%
     }

\beginsupplement



```{r message=FALSE, warning=FALSE, echo = FALSE}
library(ggmsa)
library(ggplot2)
library(ggtree)
library(gggenes)
library(ape)
library(Biostrings)
library(ggnewscale)
library(dplyr)
library(ggtreeExtra)
library(phangorn)
library(RColorBrewer)
library(patchwork)
library(ggplotify)
library(aplot)



protein_sequences <- system.file("extdata", "sample.fasta", package = "ggmsa")
nt_sequence <- system.file("extdata", "LeaderRepeat_All.fa", package = "ggmsa")
miRNA_sequences <- system.file("extdata", "seedSample.fa", package = "ggmsa")
tp53_sequences <-  system.file("extdata", "tp53.fa", package = "ggmsa")
tp53_genes <- system.file("extdata", "TP53_genes.xlsx", package = "ggmsa")


library(knitr)
knit_hooks$set(crop = hook_pdfcrop)
opts_chunk$set(crop = TRUE)
```


<!-- ## Fig. \ref{fig:custom}: Example of a custom color scheme -->

<!-- Customizing the color scheme is allowed. Users can create a data frame with two columns named `names` and `color`. This data frame includes residue letters and coloring code (see below).  -->

<!-- ```{r  warning = FALSE, message = FALSE} -->
<!-- library(RColorBrewer) -->
<!-- library(pals) -->
<!-- protein_sequences <- system.file("extdata", "sample.fasta", package = "ggmsa") -->
<!-- my_pal <- colorRampPalette(rev(brewer.pal(n = 9, name = "Reds"))) -->
<!-- my_cutstom <- data.frame(names = c(LETTERS[1:26],"-"),  -->
<!--                          color = my_pal(27),  -->
<!--                          stringsAsFactors = F) -->
<!-- head(my_cutstom) -->
<!-- ``` -->

<!-- ```{r custom, fig.height = 2, fig.width = 8, warning = FALSE, message = FALSE, fig.cap="\\label{fig:custom}Assigning color codes for each character, ggmsa enables users to customize color schemes. The same applies to sequence logos."} -->
<!-- ggmsa(protein_sequences, 300, 345,  -->
<!--       custom_color = my_cutstom,  -->
<!--       char_width = 0.5,  -->
<!--       border = "white") -->
<!-- ``` -->

<!-- \newpage -->

## Fig. \ref{fig:color_scheme}: Examples of amino acid and nucleotide color schemes in ggmsa
```{r color_scheme, echo=FALSE, fig.height=5, fig.width=5, fig.cap="\\label{fig:color_scheme}(A) Examples of amino acid color schemes. Schemes are either quantitative, reflecting empirical or statistical properties of amino acids; or qualitative, reflecting physicochemical attributes. Chemistry is colored according to side-chain chemistry also used in DNASTAR applications [@Burland2000DNASTAR]; Shapely matches the RasMol amino acid color schemes, which are, in turn, based on Robert Fletterick’s Shapely models. Zappo is a qualitative scheme developed by M. Clamp. The residues are colored according to their physicochemical properties; Taylor [@Taylor1997Residual] is taken from Taylor and is also used in JalView [@waterhouse2009jalview]; Hydrophobicity colors the residues in the alignment based on the hydrophobicity table [@kyte1982simple]. B, X, Z, J, O, and U are amino acid ambiguity codes: B is aspartate or asparagine; Z is glutamate or glutamine; X, J, O, U is an unknown (or ‘other’); “-” indicate a gap. (B) Examples of nucleotide color schemes used by ggmsa."}

library(patchwork)
scheme_AA <- ggmsa:::scheme_AA
scheme_NT <- ggmsa:::scheme_NT

scheme_AA <- scheme_AA[, - 6]
scheme_AA$char <- row.names(scheme_AA)

order_char <- c("A","R", "N", "D", "C", "Q", "E", "G", "H", "I", "L", "K", "M",
                "F","P", "S", "T", "W", "Y", "V", "B", "X", "Z", "J", "O", "U", "-")

scheme_AA$char <- factor(scheme_AA$char, levels = rev(order_char))
my_scheme <- tidyr::gather(scheme_AA, key = scheme, value = color, -char)

my_scheme$scheme <- factor(my_scheme$scheme, levels = c("Chemistry_AA","Shapely_AA","Zappo_AA","Taylor_AA","Hydrophobicity" ,"LETTER" ))
my_scheme[my_scheme$color == "","color"] <- NA

p1_sch <- ggplot(data = my_scheme, aes_(x =~ scheme, y =~ char, fill =~ I(color))) + geom_tile(color = "#f5f5f5", size = 2) +
    ylab("Amino acid") + xlab(NULL) +
    scale_x_discrete(position = "top") +
    theme(axis.ticks = element_blank(),
          axis.text.x = element_text(angle = 45, hjust = 0),
          panel.background = element_blank()) +
    coord_fixed(ratio = 0.7)

scheme_NT$char <- row.names(scheme_NT)
order_charNT <- c("A", "C", "G", "T", "U", "-")
scheme_NT$char <- factor(scheme_NT$char, levels = rev(order_charNT))
my_schemeNT <- tidyr::gather(scheme_NT, key = scheme, value = color, -char)

#########################NT colors##################
p2_sch <- ggplot(data = my_schemeNT, aes_(x =~ scheme, y =~ char, fill =~ I(color))) + geom_tile(color = "#f5f5f5", size = 2) +
  ylab("Nucleotide") + xlab(NULL) +
  scale_x_discrete(position = "top") +
  theme(axis.ticks = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 0),
        panel.background = element_blank()) +
  coord_fixed(ratio = c(1.5))


(p1_sch | (p2_sch / plot_spacer()))+ plot_annotation(tag_levels = 'A')
```

\newpage



<!-- ## Fig. \ref{fig:birds_eye}: Bird's-eyes view in ggmsa -->

<!-- ```{r birds_eye, fig.height = 1.5,fig.width = 11,  message=F, fig.cap="\\label{fig:birds_eye}A compact overview MSA showing a bird’s eye view of the full alignment of  the Cripavirus Internal Ribosomal Entry Site [family RF00458 from the Rfam database][@nawrocki2015rfam]."} -->
<!-- RF00458_msa<- system.file("extdata", "Rfam", "RF00458.fasta", package = "ggmsa") -->
<!-- ggmsa(RF00458_msa, font = NULL, border = NA, seq_name = F, color = "Chemistry_NT") +  -->
<!--   coord_cartesian() -->
<!-- ``` -->


<!-- ## Fig. \ref{fig:by_conse}: The shading methods: by_conservation  -->
<!-- ```{r by_conse, fig.height = 2, fig.width = 8, warning = FALSE, fig.cap="\\label{fig:by_conse}Example of 'by_conservation' view for MSA. This example shows the results of first coloring by 'hydrophobicity' then performing conservation analysis. Sequence alignment has the most conserved near 330 sites."} -->
<!-- ggmsa(protein_sequences, 300, 350, color = "Hydrophobicity",  -->
<!--       font = NULL, seq_name = T ,border = "white", by_conservation = T) -->
<!-- ``` -->


<!-- ## Fig. \ref{fig:logo}: Protein sequence logos annotation for an MSA -->

<!-- The `+` is used to link main function `ggmsa()` and `geom_seqlogo()` function. The `geom_seqlogo()` generates sequence logos according to each column’s molecular frequency distribution on MSA.  -->
<!-- ```{r logo, fig.height = 2, fig.width = 8, warning = FALSE, fig.cap="\\label{fig:logo}Sequence Logos between 300 and 350 sites of PH4H which play a major role in indentifying DNA, RNA, and protein binding sites [@schneider1990sequence]  "} -->
<!-- ggmsa(protein_sequences, 300, 350, char_width = 0.5, seq_name = T ) +  -->
<!--     geom_seqlogo(color = "Chemistry_AA") -->
<!-- ``` -->



<!-- ## Fig. \ref{fig:gc}: DNA sequence logos and GC bubbles for an MSA -->

<!-- `geom_GC()` calculating the GC-content for each DNA/RNA sequence. The sizes of bubbles are identical with GC-content. -->
<!-- ```{r gc, fig.height = 4, fig.width = 11, warning = FALSE, fig.cap="\\label{fig:gc}Example of DNA sequence logos (top) and GC-content (right) automatically generated by ggmsa.The size of the bubbles determined the GC content of each sequence."} -->
<!-- nt_sequence <- system.file("extdata", "LeaderRepeat_All.fa", package = "ggmsa") -->
<!-- ggmsa(nt_sequence,font = NULL, color = "Chemistry_NT") +  -->
<!--     geom_seqlogo(color = "Chemistry_NT") + geom_GC() + theme(legend.position = "none") -->
<!-- ``` -->


## Fig. \ref{fig:seed1}: miRNA sequence seed region annotation for MSA (asterisks)

`geom_seed()` helps to identify microRNA seed region by asterisks or shaded area. The seed region is a conserved heptameric sequence that is mostly situated at positions 2-7 from the miRNA 5´-end.
```{r seed1, fig.height = 2, fig.width =6, warning = FALSE, message=FALSE, fig.cap="\\label{fig:seed1}Example of highlighting miRNA seed region. Asterisks are used for marking the seed region. "}
miRNA_sequences <- system.file("extdata", "seedSample.fa", package = "ggmsa")
ggmsa(miRNA_sequences, char_width = 0.5, color="Chemistry_NT") + 
    geom_seed(seed = "GAGGUAG", star = TRUE) + coord_cartesian()
```



## Fig. \ref{fig:seed2}: miRNA sequence seed region annotation for MSA (The shaded block)

```{r seed2, fig.height = 2, fig.width = 6, warning = FALSE,message=FALSE, fig.cap="\\label{fig:seed2}Example of highlighting miRNA seed region. Annotation of sequence seed region with the shaded block"}
ggmsa(miRNA_sequences, char_width = 0.5, seq_name = T, none_bg = TRUE) + 
    geom_seed(seed = "GAGGUAG") + coord_cartesian()
```

\newpage

## Fig. \ref{fig:conse}: The consensus sequence

```{r conse, fig.height = 2, fig.width = 8, warning = FALSE, fig.cap="\\label{fig:conse}Example of consensus sequence for an MSA. The consensus sequence is displayed above the alignment and shows which residues are conserved, which residues are variable. A consensus is constructed from the most frequent residues at each site."}
ggmsa(protein_sequences, 300, 350, char_width = 0.5, 
      seq_name = T, consensus_views = T, use_dot = T)
```



<!-- ## Fig. \ref{fig:gghelix}: RNA secondary structure visulization using ggmsa -->

<!-- ggmsa supports plotting RNA secondary structure as arc diagram by reference to R4RNA [@2012R]. The ‘overlapping structure diagram’ helps to  compare RNA secondary structure between R FAM database (known) and base-pair predicted by T RANSAT [@wiebe2010transat]. The structure shown above the horizontal sequence is the known structure, colored by P-value if correctly predicted by T RANSAT (best in dark blue and worst in light blue). -->

<!-- ```{r gghelix, fig.height = 3, fig.width = 6, warning = FALSE, fig.cap="\\label{fig:gghelix}An example of an overlapping structure diagram, showing the Cripavirus Internal Ribosomal Entry Site. The structure shown above the horizontal sequence is known and  the structure below  is predicted by T RANSAT"} -->
<!-- transat_file <- system.file("extdata", "helix.txt", package = "R4RNA") -->
<!-- known_file <- system.file("extdata", "vienna.txt", package = "R4RNA") -->

<!-- known <- readSSfile(known_file, type = "Vienna" ) -->
<!-- transat <- readSSfile(transat_file, type = "Helix" ) -->
<!-- #gghelix(known) -->
<!-- gghelix(list(known = known, predicted = transat), color_by = "value", overlap = T) -->
<!-- ``` -->

<!-- \newpage -->

## Fig. \ref{fig:facet}: MSA view with break down layout

`facet_msa()` module allows to showing more alignment data in a restricted canvas. The long sequence was broken down and displayed in several lines.
```{r facet, fig.height = 6, fig.width = 10, message=FALSE, warning=FALSE, fig.cap="\\label{fig:facet} The long sequence was broken down and displayed in several lines."}
 # 4 fields
 ggmsa(protein_sequences, start = 0, end = 400, font = NULL, color = "Chemistry_AA") + 
    facet_msa(field = 100)
```


## Fig. \ref{fig:treeExtra}: MSA view with circular layout tree

A specific layout of the alignment can also be displayed by linking ggtreeExtra [@xu2021ggtreeextra]. `geom_fruit` will automatically align MSA graphs to the tree with circular layout 
```{r treeExtra, fig.height=3, fig.width =3, message = FALSE, warning= FALSE, fig.cap="\\label{fig:treeExtra}Example of MSA view with circular layout phylogenetic tree."}
library(ggtree)
library(ggtreeExtra)
sequences <- system.file("extdata", "sequence-link-tree.fasta", package = "ggmsa")

x <- readAAStringSet(sequences)
d <- as.dist(stringDist(x, method = "hamming")/width(x)[1])
tree <- bionj(d)
data <- tidy_msa(x, 120, 200)

p1 <- ggtree(tree, layout = 'circular') + 
    geom_tiplab(align = TRUE, offset = 0.545, size = 2) + 
        xlim(NA, 1.3)
p1 + geom_fruit(data = data, geom = geom_msa, offset = 0, 
                pwidth = 1.2, font = NULL, border = NA)
```

\newpage


## Fig. \ref{fig:tree}: Visualizing MSA with phylogenetic tree

ggmsa supports to link ggtree [@yu2017ggtree] by `geom_facet()`. Sequence order will automatically align with left nodes.
```{r tree, fig.height = 3, fig.width = 11, warning = FALSE, fig.cap="\\label{fig:tree}A tree resulting from the application of phylogenetic analysis to the alignment."}
 x <- readAAStringSet(protein_sequences)
 d <- as.dist(stringDist(x, method = "hamming")/width(x)[1])
 library(ape)
 tree <- bionj(d)
 library(ggtree)
 p <- ggtree(tree) + geom_tiplab()

 data = tidy_msa(x)
 p + geom_facet(geom = geom_msa, data = data,  panel = 'msa',
                font = NULL, border = NA, color = "Chemistry_AA") +
     xlim_tree(1)
```

## Fig. \ref{fig:re-order_bundles}: Re-designated residues order in Sequence bundle

```{r re-order_bundles, fig.height = 4, fig.width = 11, warning = FALSE, fig.cap="\\label{fig:re-order_bundles}The preference of residues’ physicochemical property can be reflected by adjusting the order of letters on the y-axis. It shows re-dsignaed residues order according to amino acid molecule weight. The large molecule is at the bottom"}
negative <-  system.file("extdata", "Gram-negative_AKL.fasta", 
                         package = "ggmsa")
positive <-  system.file("extdata", "Gram-positive_AKL.fasta", 
                         package = "ggmsa")

ggSeqBundle(list(negative, positive), 
                 alpha = 0.1, 
                 bundle_color = c("#FC8D62","#8DA0CB"),
                 lev_molecule = c("-", "W", "Y", "R", "F", "H", "M", "E", 
                                  "K", "Q", "D", "N", "L", "I", "C", "T", 
                                  "V", "P", "S", "A", "G"))
```


\newpage

## Table S1: Comparison of ggmsa with popular free MSA visualization tools

We compared MSA visualization methods among ggmsa and other tools (**msa** [@bodenhofer2015msa], **MSAviewer** [@yachdav2016msaviewer], **AliView** [@larsson2014aliview], **Jalview** [@waterhouse2009jalview], **ALVIS** [@schwarz2016alvis])

```{r compare, echo=FALSE, message=FALSE, warning=FALSE,  fig.width=7.5, fig.height=3.2, error=FALSE, fig.cap="\\label{fig:compare}Visualization methods for MSA supported by different software tools, results='asis'."}
require(kableExtra)

x <- "ggmsa\tR package\tYES\tYES\tYES\trectangular fragmentary circular\tYES\tYES\tprogramming and command line\n
msa\tR package\tYES\tNO\tYES\trectangular\tNO\tNO\tprogramming and command line\n
MSAviewer\tWeb service\tYES\tNO\tYES\trectangular\tNO\tNO\tinteractive\n
AliView\tDesktop application\tNO\tNO\tYES\trectangular\tNO\tNO\tinteractive\n
Jalview\tDesktop application and web service\tYES\tNO\tYES\trectangular\tNO\tNO\tinteractive\n
ALVIS\tDesktop application\tYES\tYES\tYES\trectangular\tNO\tNO\tinteractive\n
"

caption <- "Comparison of ggmsa with popular free MSA visualization tools"

header <- c("Tools","Platform","Sequence Logos","Sequence bundles", 
            "Stacked MSA[note]", "Layouts for stacked MSA", 
            "Integrating external data into stacked MSA[note]",
            "Exploring sequence recombination[note]", "User interface")

# header <- gsub(" ", "\\ ", header) %>%
#           linebreak(align="l", linebreaker = ",")

 y <-  read.table(text=x, header=F, sep="\t", quote="", check.names=F) 
compare_tab <-
   y %>%
  knitr::kable(
               caption=caption,
               col.names=header,
               booktabs = T,
               position = "!h",
               escape=F,
               align = "c"
  ) %>%
  row_spec(0, bold=T) %>%
  row_spec(row=c(1,2,3,4,5), hline_after=T) %>%
  kable_styling(
                latex_options = c("striped"),
                full_width = T,
                font_size = 6.6,
                stripe_color = "gray!10",
                position="center"
  )%>%
  add_footnote(c(
                 "Stacked MSA: it represents all sequences as rows and homologous residue positions as columns",
                 "Extended MSA: adding associated into stacked MSA plots",
                 "A visualization method that designed for detecting sequence recombination signals"
                 ),
               notation="number")%>%
  column_spec(1, width = "5em", latex_valign = "m") %>%
  column_spec(2, width = "6em", latex_valign = "m")%>%
  column_spec(3, width = "5em", latex_valign = "m")%>%
  column_spec(4, width = "7em", latex_valign = "m")%>%
  column_spec(5, width = "6em", latex_valign = "m")%>%
  column_spec(6, width = "5em", latex_valign = "m")%>%
  column_spec(7, width = "5em", latex_valign = "m")%>%
  column_spec(8, width = "6em", latex_valign = "m")%>%
  column_spec(9, width = "5em", latex_valign = "m")
compare_tab
```
<!-- \newpage -->


## Fig. 2A: The combination of sequence logos and sequence bundles (R code)

```{r, fig.height=3, fig.width=8, message=FALSE}
p2a <- seqlogo("data/Gram-NP-merge.fa", 
               color = "Chemistry_AA", 
               font = "DroidSansMono") + 
   coord_cartesian()


negative <-  system.file("extdata", "Gram-negative_AKL.fasta", 
                         package = "ggmsa")
positive <-  system.file("extdata", "Gram-positive_AKL.fasta", 
                         package = "ggmsa")

pos <- data.frame(x= c(4, 7, 9, 24, 27, 29,
                       4, 7, 24, 27), 
                  y = c(c(21, 11, 20, 17, 12, 18) +.3,
                        c(13, 13, 13, 13) +.5
                        ), 
                  label = c("H", "S", "R", "D", "T", "E",
                            "C", "C", "C", "C"),
                  color = c(rep("#ff4700",6), 
                            rep("#0443d0",4)))

p2b <- ggSeqBundle(list(negative, positive), 
                   alpha = 0.1, 
                   bundle_color = c("#FC8D62","#8DA0CB"))+ #RColorBrewer: Set2:2-3
  geom_text(data = pos, 
            mapping = aes(x, y, 
                          label = label, 
                          color = I(color)),
            inherit.aes = FALSE, 
            size = 4)

plot_list(gglist = list(p2a, p2b), ncol = 1, heights = c(0.3,1))
```
**Fig. 2A**| The combination of sequence logos and sequence bundles. The data contain adenylate kinase lid (AKL) domain both of Gram-negative and Gram-negative bacteria. 100 sequences for each groups. The sequence logo (top panel) represents the AKL sequence pattern and the sequence bundle (bottom panel) represents the different residues relationship between Gram-negative (orange) and Gram-negative (purple) bacteria. The site at 4, 7, 9, 24, 27 and 29 has exclusive pattern (His4, Ser7, Arg9, Asp24, Thr27, Glu29) in the Gram-negative sequences. And the site at 4, 7, 24 and 27 both contain the Cysteines in the Gram-positive sequences. These residues relationship in agreement with the structural stability with the AKL domain. Gram-negatives form hydrogen bonding network by the exclusive pattern and Gram-positives bound metal ion to form coordinated tetrahedral by Cys. (Date from BioVis2013 and repeated example from Science Practice)



##  Fig. 2B: An example of MSA visualization and MSA annotations (R code)
```{r fig.height = 3, fig.width = 11, warning = FALSE, message = FALSE}
 ggmsa(protein_sequences, 221, 280, seq_name = TRUE, char_width = 0.5) +
    geom_seqlogo(color = "Chemistry_AA") +
        geom_msaBar()
```
**Fig. 2B**|A local visualization of the sequence alignment of the phenylalanine hydroxylase protein (PH4H) within nine species. The center panel is the main MSA plot with residues in the alignment colored according to the Chemistry color scheme (amino acids are colored according to their sidechain chemistry). The top and bottom panels are corresponding annotations with MSAs, showing the conservation patterns at each position by sequence logos and the distribution of the high-frequency residue by a bar chart, respectively. 

#&nbsp; 
\newpage

## Fig. 2CD: Visual methods of surveying RNA co-variation and structural changes (R code)

```{r  fig.height = 11, fig.width = 10, warning=FALSE}
RNA7S  <- "data/3JAJ-2D-dotbracket.txt"
RNAP54 <- "data/4UJE-2D-dotbracket.txt"

RF03120_msa<- system.file("extdata", "Rfam", "RF03120.fasta", package = "ggmsa")
RF03120_ss <- system.file("extdata", "Rfam", "RF03120_SS.txt", package = "ggmsa")

known <- readSSfile(RNA7S, type = "Vienna" )
transat <- readSSfile(RNAP54 , type = "Vienna")


RF_arc <- readSSfile(RF03120_ss, type = "Vienna" )

p2C <- ggmsa(RF03120_msa, 
             font = NULL, 
             color = "Chemistry_NT", 
             seq_name = F, 
             show.legend = F, 
             border = NA) +
        geom_helix(helix_data = RF_arc) + 
        theme(axis.text.y = element_blank())

p2D <- ggmsa("data/5SRNA.fa",
             font = NULL,
             color = "Chemistry_NT",
             seq_name = T,
             show.legend = T,
             border = NA) +
  geom_helix(helix_data = list(known = known, 
                               predicted = transat),
             overlap = F)

p2CD <- plot_list(gglist = list(p2C, p2D), 
                  ncol = 1,heights = c(0.15), 
                  tag_levels = list(c("2C","2D" )))
 
p2CD
```
**Fig. 2CD**| **(C)** The data from the Rfam database [family RF03120] include 19 seed alignments of Sarbecovirus 5'UTR (including 6 SARS-CoV-2 isolates sequences) and the corresponding consensus RNA secondary structure. Compensatory mutations in MSA can be detected by checking alignment columns in positions corresponding to arcs. **(D)** The sequence and secondary structure of 5S Ribosomal RNA from the PDB database. The arc above the sequence represents the structure of the engaged state of the mammalian SRP-ribosome complex (3JAJ) and the bottom arc depicts the 5S RNA structural changes in a eukaryotic-specific ribosome rearrangement (4UJE).

## Fig. 3: Visual exploration for sequence recombination signal (R code)

```{r  fig.height = 11, fig.width = 10, warning=FALSE, message=FALSE}
fas <- c("data/HM_KP.fa","data/CK_KP.fa")
xx <- lapply(fas, seqdiff)
plts <- lapply(xx, plot, width = 100)
plts[[3]] <- simplot("data/CK_HM_KP.fa", 'KP827649', smooth = FALSE) + 
  theme(legend.position = "bottom")

plot_list(gglist=plts, ncol=1, tag_levels = list(c("A",' ',"B", ' ',"C")))
```
**Fig. 3**| The nucleotide different plots were generated using the comparison of KP827649 sequence (Tomato Spotted Wilt Virus large RNA genomic segment) with that of (A) HM581979 sequence (Major parent) and (B) KC261971 sequence (Minor parent). (C) The similarity curves investigates the sequence similarity between KP827649 sequence and potential parents. Two recombination signals (The intersection of two curves in similarity plot, start breakpoint is 4534 and the end breakpoint is 5536) were detected in the TSWV LRNA genomic segment. 


## Fig. 4A: Examples of graphics combination (R code)

```{r  fig.height = 4, fig.width = 15, warning=FALSE}
##Fig 6A tree + msa + genes locus
dat <- read.aa(tp53_sequences, format = "fasta") %>% phyDat(type = "AA", levels = NULL)
tree <- dist.ml(dat, model = "JTT") %>% bionj()
dd <- ggimage::phylopic_uid(tree$tip.label)

p_tp53 <- ggtree(tree, branch.length = 'none') %<+% dd +
  geom_tiplab(aes(image=uid), geom = "phylopic", offset =1.9) +
  geom_tiplab(aes(label=label)) +
  geom_treescale(x = 0,y = -1)
#msa
data_53 <- readAAMultipleAlignment(tp53_sequences) %>% tidy_msa()
#gene maps
TP53_arrow <- readxl::read_xlsx(tp53_genes)
TP53_arrow$direction <- 1
TP53_arrow[TP53_arrow$strand == "reverse","direction"] <- -1

#color
mapping = aes(xmin = start, xmax = end, fill = gene, forward = direction)
my_pal <- colorRampPalette(rev(brewer.pal(n = 10, name = "Set3")))

#tree + gene maps + msa
p4a <- p_tp53  + xlim_tree(4) +
  geom_facet(geom = geom_msa, data = data_53,
             panel = 'Multiple Sequence Alignment of the TP53 Protein', font = NULL,
             border = NA) +
  new_scale_fill() +
  scale_fill_manual(values = my_pal(10)) +
  geom_facet(geom = geom_motif,
             mapping = mapping, data = TP53_arrow,
             panel = 'Genome_Locus',  on = 'TP53',
             arrowhead_height = unit(3, "mm"),
             arrowhead_width = unit(1, "mm")) +
  theme(strip.background=element_blank(),
        strip.text = element_text(size = 13))
p4A <- facet_widths(p4a, c(Tree = 0.35, Genome_Locus = 0.3))
p4A
```
**Fig. 4A**| The MSA-tree panel in conjunction with external genome locus data set. Comparative genome locus structure (genome_Locus panel), sequence alignment of TP53 protein (the middle panel), and the corresponding phylogenetic tree (Tree panel) among six species. The local genome map shows the 30000 sites around the TP53 gene, and the phylogenetic tree that represents evolutionary relationships inferred from TP53 protein sequences using the Neighbor-Joining method based on the evolutionary distances of JTT matrix-based method.


## Fig. 4B: Examples of graphics combination (R code)

```{r  fig.height = 13, fig.width = 16, warning=FALSE}
##Fig 6B tree + msa + 2boxplot
seq <-readDNAStringSet("data//btuR.fa")
aln <- tidy_msa(seq)
btuR_tree <- read.tree("data/btuR.nwk")
meta_dat <- read.csv("data/meta_data_47.csv")

#Pathotype_fill_colors
Pathotype_cols <- RColorBrewer::brewer.pal(7, "Set3")
names(Pathotype_cols) <- meta_dat$Pathotypes %>% factor %>% levels

###tree OTU
Phylo_group <- list(A= meta_dat$Lineage[meta_dat$Phylogroup == "A"]%>% unique,
                    B1=meta_dat$Lineage[meta_dat$Phylogroup == "B1"]%>% unique,
                    B2=meta_dat$Lineage[meta_dat$Phylogroup == "B2"]%>% unique,
                    C=meta_dat$Lineage[meta_dat$Phylogroup == "C"]%>% unique,
                    D =meta_dat$Lineage[meta_dat$Phylogroup == "D"]%>% unique,
                    E =meta_dat$Lineage[meta_dat$Phylogroup == "E"]%>% unique,
                    `F`=meta_dat$Lineage[meta_dat$Phylogroup == "F"]%>% unique,
                    Shigella=meta_dat$Lineage[meta_dat$Phylogroup == "Shigella"]%>% unique)

Phylo_cols <- RColorBrewer::brewer.pal(8, "Dark2")
names(Phylo_cols) <- names(Phylo_group)

## plot tree
p_btuR_tree <- ggtree(btuR_tree) + geom_tiplab(align = T)
p_btuR_tree <- groupOTU(p_btuR_tree ,Phylo_group)+aes(color=group)  +
  scale_color_manual(values = c(Phylo_cols, "black"), na.value = "black", name = "Lineage",
                     breaks = c("A", "B1", "B2", "C", "D", "E", "F", "Shigella"),  guide="none")

p_btuR_tree <- p_btuR_tree +
  geom_strip('L29', 'L20', barsize=2, color=Phylo_cols[["B2"]],
             label="B2",offset =.01,  offset.text = 0.0015) +
  geom_strip('L28','L29', barsize=2, color=Phylo_cols[["A"]],
             label="A",offset =.01,  offset.text = 0.0015) +
  geom_strip('L15','L28', barsize=2, color=Phylo_cols[["B1"]],
             label="B1",offset =.01,  offset.text = 0.0015) +
  geom_strip('L45','L15', barsize=2, color=Phylo_cols[["Shigella"]],
             label="S.",offset =.01,  offset.text = 0.0015) +
  geom_strip('L36','L45', barsize=2, color=Phylo_cols[["B1"]],
             label="B1",offset =.01,  offset.text = 0.0015) +
  geom_strip('L30','L36', barsize=2, color=Phylo_cols[["Shigella"]],
             label="S.",offset =.01,  offset.text = 0.0015) +
  geom_strip('L39','L30', barsize=2, color=Phylo_cols[["B1"]],
             label="B1",offset =.01,  offset.text = 0.0015) +
  geom_strip('L40','L39', barsize=2, color=Phylo_cols[["C"]],
             label="C",offset =.01,  offset.text = 0.0015)  +
  geom_strip('L48','L40', barsize=2, color=Phylo_cols[["B1"]],
             label="B1",offset =.01,  offset.text = 0.0015) +
  geom_strip('L10','L48', barsize=2, color=Phylo_cols[["E"]],
             label="E",offset =.01,  offset.text = 0.0015) +
  geom_strip('L37','L10', barsize=2, color=Phylo_cols[["D"]],
             label="D",offset =.01,  offset.text = 0.0015)+
  geom_strip('L33','L37', barsize=2, color=Phylo_cols[["F"]],
             label="F",offset =.01,  offset.text = 0.0015)+
  geom_strip('L1','L33', barsize=2, color=Phylo_cols[["E"]],
             label="E",offset =.01,  offset.text = 0.0015)

##tree + meta data boxplots
p4B <- p_btuR_tree  +
  geom_treescale(x = 0,y = -1) +
  geom_fruit(data = aln,
             geom = geom_msa,
             end = 200,
             font = NULL,
             color = "Chemistry_NT",
             border = NA,
             consensus_views = T, 
             ref = "L38",
             pwidth = 3.5,
             offset = 0.3,
             axis.params = list(title = "Multiple Sequence Alignment of the btuR Gene",
                                title.height = 0.05,
                                title.size = 4.5,
                                axis = "x",
                                vjust = 1.1,
                                text.size = 3,
                                line.size = 1,
                                line.color = "black")) +
  new_scale_fill() +
  geom_fruit(mapping = aes(x = AMR_genes, y = Lineage, fill = MDR),
             data = meta_dat,
             geom = geom_boxplot,
             outlier.size = 0.5,
             pwidth=1,
             offset = 0.1,
             axis.params = list(title = "Antimicrobial Classes",
                                title.height = 0.05,
                                title.size = 4.5,
                                axis = "x",
                                vjust = 1.1,
                                text.size = 3,
                                line.size = 1,
                                line.color = "black")) +
  scale_fill_manual(values=c("Yes" = "#fcd7c5", "No" = "#dfdfdf")) +
  new_scale_fill() +
  geom_fruit(mapping = aes(x = virulence_genes, y = Lineage, fill = Pathotypes),
             data = meta_dat,
             geom = geom_boxplot,
             pwidth=1,
             offset = 0.05,
             axis.params = list(title = "Virulence Genes",
                                title.height = 0.05,
                                title.size = 4.5,
                                axis = "x",
                                vjust = 1.1,
                                text.size = 3,
                                line.size = 1,
                                line.color = "black"))+
  scale_fill_manual(values = Pathotype_cols)

p4B
```
**Fig. 4B**| The summarized visualization for different E.coli lineages. The alignment of btuR gene was selected to generate a phylogenetic tree which was stained according to the corresponding lineages. The boxplots represent antimicrobial resistance (AMR) and virulence profiles of the lineages. Number of antimicrobial classes and virulence genes per isolate colored by multidrug-resistant (MDR) classes and the most prevalent predicted pathotype in the lineage.


## Session Info

 Here is the output of sessionInfo() on the system on which this document was compiled:
```{r echo = FALSE}
sessionInfo()
```


&nbsp;


## References


<!-- ##  Fig. 3B: The external data genome locus in conjunction with the MSA-tree plot (R code)  -->
<!-- ```{r fig.height=4, fig.width=14, message=FALSE, warning=FALSE} -->
<!-- library(phangorn) -->
<!-- #tree -->
<!-- dat <- read.aa(tp53_sequences, format = "fasta") %>% phyDat(type = "AA", levels = NULL) -->
<!-- tree <- dist.ml(dat, model = "JTT") %>% bionj() -->
<!-- dd <- ggimage::phylopic_uid(tree$tip.label) -->

<!-- p_tp53 <- ggtree(tree, branch.length = 'none') %<+% dd +  -->
<!--   geom_tiplab(aes(image=uid), geom = "phylopic", offset =1.9) + -->
<!--   geom_tiplab(aes(label=label))  -->
<!-- #msa -->
<!-- data_53 <- readAAMultipleAlignment(tp53_sequences) %>% tidy_msa() -->
<!-- #gene maps -->
<!-- TP53_arrow <- readxl::read_xlsx(tp53_genes) -->
<!-- TP53_arrow$direction <- 1 -->
<!-- TP53_arrow[TP53_arrow$strand == "reverse","direction"] <- -1 -->

<!-- #color -->
<!-- mapping = aes(xmin = start, xmax = end, fill = gene, forward = direction) -->
<!-- my_pal <- colorRampPalette(rev(brewer.pal(n = 10, name = "Set3"))) -->

<!-- #tree + gene maps + msa -->
<!-- pb <- p_tp53  + xlim_tree(4) + -->
<!--     geom_facet(geom = geom_msa, data = data_53, -->
<!--                panel = 'msa', font = NULL, -->
<!--                   border = NA) + -->
<!--     new_scale_fill() + -->
<!--     scale_fill_manual(values = my_pal(10)) + -->
<!--     geom_facet(geom = geom_motif, -->
<!--                mapping = mapping, data = TP53_arrow, -->
<!--                panel = 'genes',  on = 'TP53', -->
<!--                arrowhead_height = unit(3, "mm"), -->
<!--                arrowhead_width = unit(1, "mm")) -->

<!-- facet_widths(pb, c(Tree = 0.35, genes = 0.3)) -->
<!-- ``` -->
<!-- **Fig. 3B**|The external data, genome locus, in conjunction with the MSA-tree plot. Comparative genome locus structure (genes panel), sequence alignment of TP53 protein (msa panel) and the corresponding phylogenetic tree (Tree panel) among six organisms. The local genome map shows the 30000 sites around the TP53 gene, and the phylogenetic tree that represents evolutionary relationships of TP53 protein was inferred using the Neighbor-Joining method based on the evolutionary distances of JTT matrix-based method. The Danio rerio shows strong inconsistent in the alignment and remote evolutionary relationships in the tree comparing other organisms. And the inconsistent is also discovered in local genome map panel (e.g. linkage disequilibrium of TP53-WRAP53 locus deletions, the direction of transcription and the variety of neighboring genes). This combined visualization interprets the variant concordance of the TP53 molecule between molecular sequences and genome locus structures. -->





<!-- ##  Fig. 4C: RNA secondary structure annotation for a MSA (R code)  -->

<!-- ```{r fig.height = 3, fig.width = 8, message=F, warning=F} -->
<!-- RF03120_msa<- system.file("extdata", "Rfam", "RF03120.fasta", package = "ggmsa") -->
<!-- RF03120_ss <- system.file("extdata", "Rfam", "RF03120_SS.txt", package = "ggmsa") -->

<!-- known <- readSSfile(RF03120_ss, type = "Vienna" ) -->

<!-- ggmsa(RF03120_msa,  -->
<!--       font = NULL,  -->
<!--       color = "Chemistry_NT",  -->
<!--       seq_name = F,  -->
<!--       show.legend = T,  -->
<!--       border = NA) + -->
<!--   geom_helix(helix_data = known) +  -->
<!--   theme(axis.text.y = element_blank()) -->
<!-- ``` -->
<!-- **Fig. 4C**|An example of visualizing MSAs in conjunction with RNA secondary structure. The data from the Rfam database [family RF03120] include 19 seed alignments of Sarbecovirus 5'UTR (including 6 SARS-CoV-2 isolates sequences) and the corresponding consensus RNA secondary structure. The extra RNA secondary structure data integrated with MSA plot as arc diagrams and helices were colored according to the base-pair group. The 5′ UTR within the Sarbecovirus is responsible for important biological functions, such as viral replication, transcription and packaging. And it has a conserved RNA secondary structure on common Coronavirus genera. The graphical combination of the alignment and secondary structures reveals the co-variation of Sarbecovirus 5'UTR that retains the base-pairing ability, but changes the base-pairing nucleotides, and is strong evidence for RNA structure conservation. -->

<!-- ##  Fig. 5: Examples of detecting sequence recombination signals (R code)  -->

<!-- ```{r fig.height = 10, fig.width = 10, message=F, warning=F} -->
<!-- library(aplot) -->
<!-- library(ggplotify) -->
<!-- library(patchwork) -->

<!-- fas <- list.files(system.file("extdata", "GVariation", package="ggmsa"),  -->
<!--                   pattern="fas", full.names=TRUE) -->
<!-- fas <- fas[-3] -->

<!-- xx <- lapply(fas, seqdiff) -->
<!-- plts <- lapply(xx, plot) -->
<!-- plts[[3]] <- simplot(fas, 'CF_YL21') -->

<!-- plot_list(lapply(plts, function(i)as.ggplot(i)), ncol = 1) +  -->
<!--   plot_annotation(tag_levels = "A") -->
<!-- ``` -->
<!-- **Fig. 5**| Examples of detecting sequence recombination signals. CF_YL21 is a Potato virus Y (PVY) isolate from Solanum tuberosum in China. PVYN (Mont, AY884983) and PVYO (Oz, EF026074) were chosen as potential parents. (A) The comparison of CF_YL21 sequence with Mont. The x-axis corresponds to nucleotide position along the CF_YL21 genome, and the y-axis is the number of nucleotide differences between the genomes. Bar charts indicate number of base differences in every 50 sites (B) The comparison of CF_YL21 sequence with Oz (C) Detection and verification of recombination breakpoints in PVY CF_YL21 using PVYN and PVYO as reference strains. Two recombination signals (The intersection of line, site of 309 and 2224) were detected in the CF_YL21 genome. -->

<!-- &nbsp; -->
<!-- &nbsp; -->



